File: messaging/GLPImageOperation.m
//
//  GLPImageOperation.m
//  Gleepost
//
//  Created by Silouanos on 19/12/2013.
//  Copyright (c) 2013 Gleepost. All rights reserved.
//

#import "GLPImageOperation.h"
#import "NSNotificationCenter+Utils.h"
#import "ImageFormatterHelper.h"
#import "WebClient.h"
#include <semaphore.h>

@interface GLPImageOperation ()

@property (nonatomic, strong) UIImage *image;
@end

@implementation GLPImageOperation

@synthesize image = _image;

- (void)main {
    
    @autoreleasepool
    {

        
        NSLog(@"GLPImageOperation");
        
        
        [self startUploadingImage];
        
//        [[NSNotificationCenter defaultCenter] postNotificationNameOnMainThread:@"GLPImageUploaded" object:nil userInfo:@{@"remoteKey":@"TEST"}];

        
    }
}

-(id)initWithImage:(UIImage*)image
{
    self = [super init];
    
    if(self)
    {
        _image = image;
    }
    
    
    return self;
}




#pragma mark - Operational methods

- (void)startUploadingImage
{
//    _postImage = image;
//    _imageStatus = GLPImageStatusUploading;
    
    __block NSData *data;
    
    //NSBlockOperation *operation = [NSBlockOperation blockOperationWithBlock:^{
        UIImage *resizedImage = [ImageFormatterHelper imageWithImage:_image scaledToHeight:640];
        data = UIImagePNGRepresentation(resizedImage);
    //}];
    
    //[operation setCompletionBlock:^{
        [self uploadResizedImageWithImageData:data];
        
    //}];
    
//    [[[NSOperationQueue alloc] init] addOperation:operation];
}

- (void)uploadResizedImageWithImageData:(NSData *)imageData {
    
    __block BOOL finished = NO;
    __block NSString *imageUrlSend = nil;
    
    //Lock.
    
    if (imageData) {
        [[WebClient sharedInstance] uploadImage:imageData callback:^(BOOL success, NSString *imageUrl) {
            if (success) {
//                _imageStatus    = GLPImageStatusUploaded;
//                _imageURL       = imageUrl;
//                
//                if (_uploadContentBlock) _uploadContentBlock();
                
               // NSLog(@"Image url before notify: %@",imageUrl);
                
                finished = success;
                imageUrlSend = imageUrl;
                
                if(finished)
                {
                    [[NSNotificationCenter defaultCenter] postNotificationNameOnMainThread:@"GLPImageUploaded" object:nil userInfo:@{@"imageUrl":imageUrlSend}];
                    NSLog(@"Image url after notify: %@",imageUrlSend);

                }


                
            } else {
//                _imageStatus = GLPImageStatusFailed;
                
                NSLog(@"Error occured. Post image cannot be uploaded.");
            }
        }];
        
        //Unlock.


    }
}
@end

File: messaging/GLPImageOperation.swift
//
//  GLPImageOperation.swift
//  Gleepost
//
//  Created by Silouanos on 09/06/15.
//  Copyright (c) 2015 Gleepost. All rights reserved.
//

import Foundation
import UIKit

@objc protocol GLPImageOperationDelegate
{
    func imageUploaded(timestamp: String, image: UIImage, imageUrl: String)
}

class GLPImageOperation: NSOperation
{
    private var image: UIImage
    private let timestamp: String
    private var progress: Float
    
    var delegate: GLPImageOperationDelegate?
    
    override func main()
    {
        self.processAndUploadImage()
    }
    
    init(timestamp: String, image: UIImage)
    {
        self.timestamp = timestamp
        self.image = image
        self.progress = 0.0
        super.init()
        self.saveImageToCache()
    }
    
    
    //MARK: - Cache
    
    private func saveImageToCache()
    {
        GLPImageCacheHelper.storeImage(image, withImageUrl: timestamp)
    }
    
    private func replaceImageToCache(uploadedUrl: String)
    {
        GLPImageCacheHelper.replaceImage(image, withImageUrl: uploadedUrl, andOldImageUrl: timestamp)
    }
    
    //MARK: - Client
    
    private func processAndUploadImage()
    {
        WebClient.sharedInstance().uploadImage(resizeImageAndConvertToNSData(), callback: { (success, imageUrl) -> Void in
            
            if success
            {
                self.replaceImageToCache(imageUrl)
                self.delegate?.imageUploaded(self.timestamp, image: self.image, imageUrl: imageUrl)
            }
            else
            {
                println("GLPImageOperation error to upload image")
            }
            
        }) { (progress) -> Void in
            
            NSNotificationCenter.defaultCenter().postNotificationName(SwiftConstants.GLPNOTIFICATION_UPLOADING_IMAGE_CHANGED_STATUS_INTERNAL, object: self, userInfo: ["status" : progress, "timestamp" : self.timestamp])
        }
    }
    
    private func resizeImageAndConvertToNSData() -> NSData
    {
        self.image = ImageFormatterHelper.imageWithImage(self.image, scaledToHeight: 640.0)
        return UIImagePNGRepresentation(self.image)
    }
}
